File: src/main/java/application/gameplay/PlayerActionContext.java
package application.gameplay;

import application.game.Game;
import model.basic.Tile;
import model.players.Player;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public class PlayerActionContext {
    private final Game game;
    private final Player player;

    public PlayerActionContext(Game game, Player player) {
        this.game = game;
        this.player = player;
    }

    public boolean isActionNeeded() {
        if (player.containsHu()) {
            return true;
        }
        if (player.isPlaying()) {
            return true;
        }
        return player.containsResponseAction();
    }

    public List<String> legalActionLabels() {
        List<String> actions = new ArrayList<>();
        if (player.containsHu()) {
            actions.add("hu");
        }
        if (!player.isPlaying() && player.containsResponseAction()) {
            addResponseActions(actions);
            return actions;
        }
        if (player.isPlaying() && player.containsKong()) {
            actions.add("kong");
            actions.add("skip");
            return actions;
        }
        if (player.isPlaying()) {
            for (Tile tile : uniqueDiscardTiles()) {
                actions.add("discard:" + TileCodec.code(tile));
            }
        }
        return actions;
    }

    private void addResponseActions(List<String> actions) {
        if (player.containsChow()) {
            actions.add("chow");
        }
        if (player.containsPung()) {
            actions.add("pung");
        }
        if (player.containsKong()) {
            actions.add("kong");
        }
        actions.add("skip");
    }

    public boolean isLegal(PlayDecision decision) {
        if (decision == null) {
            return false;
        }
        switch (decision.getAction()) {
            case HU:
                return player.containsHu();
            case CHOW:
                return !player.isPlaying() && player.containsChow();
            case PUNG:
                return !player.isPlaying() && player.containsPung();
            case KONG:
                return player.containsKong();
            case SKIP:
                return (!player.isPlaying() && player.containsResponseAction())
                        || (player.isPlaying() && player.containsKong());
            case DISCARD:
                return player.isPlaying() && !player.containsChouPungKong()
                        && resolveDiscardTile(decision.getTile()) != null;
            default:
                return false;
        }
    }

    public Tile resolveDiscardTile(Tile requestedTile) {
        if (requestedTile == null) {
            return null;
        }
        Tile newTile = player.getHand().getNewTile();
        if (newTile != null && newTile.equals(requestedTile)) {
            return newTile;
        }
        for (Tile tile : player.getHand().toList()) {
            if (tile.equals(requestedTile)) {
                return tile;
            }
        }
        return null;
    }

    public List<Tile> uniqueDiscardTiles() {
        Set<String> seen = new LinkedHashSet<>();
        List<Tile> tiles = new ArrayList<>();
        Tile newTile = player.getHand().getNewTile();
        if (newTile != null && seen.add(TileCodec.code(newTile))) {
            tiles.add(newTile);
        }
        for (Tile tile : player.getHand().toList()) {
            if (seen.add(TileCodec.code(tile))) {
                tiles.add(tile);
            }
        }
        return tiles;
    }

    public Tile getLastPlayedTile() {
        return game.getLastPlayedTile();
    }
}

File: UnityVersion/Assets/Scripts/Core/Gameplay/PlayerActionContext.cs
using System.Collections.Generic;
using SichuanMahjong.Core.Model;

namespace SichuanMahjong.Core.Gameplay
{
    public class PlayerActionContext
    {
        private readonly Game game;
        private readonly Player player;

        public PlayerActionContext(Game game, Player player)
        {
            this.game = game;
            this.player = player;
        }

        public bool IsActionNeeded()
        {
            if (player.ContainsHu())
            {
                return true;
            }
            if (player.IsPlaying())
            {
                return true;
            }
            return player.ContainsResponseAction();
        }

        public List<string> LegalActionLabels()
        {
            List<string> actions = new List<string>();
            if (player.ContainsHu())
            {
                actions.Add("hu");
            }
            if (!player.IsPlaying() && player.ContainsResponseAction())
            {
                AddResponseActions(actions);
                return actions;
            }
            if (player.IsPlaying() && player.ContainsKong())
            {
                actions.Add("kong");
                actions.Add("skip");
                return actions;
            }
            if (player.IsPlaying())
            {
                foreach (Tile tile in UniqueDiscardTiles())
                {
                    actions.Add("discard:" + TileCodec.Code(tile));
                }
            }
            return actions;
        }

        private void AddResponseActions(List<string> actions)
        {
            if (player.ContainsChow())
            {
                actions.Add("chow");
            }
            if (player.ContainsPung())
            {
                actions.Add("pung");
            }
            if (player.ContainsKong())
            {
                actions.Add("kong");
            }
            actions.Add("skip");
        }

        public bool IsLegal(PlayDecision decision)
        {
            if (decision == null)
            {
                return false;
            }
            switch (decision.GetAction())
            {
                case PlayDecision.ActionType.HU:
                    return player.ContainsHu();
                case PlayDecision.ActionType.CHOW:
                    return !player.IsPlaying() && player.ContainsChow();
                case PlayDecision.ActionType.PUNG:
                    return !player.IsPlaying() && player.ContainsPung();
                case PlayDecision.ActionType.KONG:
                    return player.ContainsKong();
                case PlayDecision.ActionType.SKIP:
                    return (!player.IsPlaying() && player.ContainsResponseAction())
                           || (player.IsPlaying() && player.ContainsKong());
                case PlayDecision.ActionType.DISCARD:
                    return player.IsPlaying() && !player.ContainsChouPungKong()
                           && ResolveDiscardTile(decision.GetTile()) != null;
                default:
                    return false;
            }
        }

        public Tile ResolveDiscardTile(Tile requestedTile)
        {
            if (requestedTile == null)
            {
                return null;
            }
            Tile newTile = player.GetHand().GetNewTile();
            if (newTile != null && newTile.Equals(requestedTile))
            {
                return newTile;
            }
            foreach (Tile tile in player.GetHand().ToList())
            {
                if (tile.Equals(requestedTile))
                {
                    return tile;
                }
            }
            return null;
        }

        public List<Tile> UniqueDiscardTiles()
        {
            HashSet<string> seen = new HashSet<string>();
            List<Tile> tiles = new List<Tile>();
            Tile newTile = player.GetHand().GetNewTile();
            if (newTile != null && seen.Add(TileCodec.Code(newTile)))
            {
                tiles.Add(newTile);
            }
            foreach (Tile tile in player.GetHand().ToList())
            {
                if (seen.Add(TileCodec.Code(tile)))
                {
                    tiles.Add(tile);
                }
            }
            return tiles;
        }

        public Tile GetLastPlayedTile()
        {
            return game.GetLastPlayedTile();
        }
    }
}
